File: API/2316_Polarized_Fractal_Efficiency/CS/PolarizedFractalEfficiencyStrategy.cs
using System;
using System.Collections.Generic;

using Ecng.Common;

using StockSharp.Algo.Indicators;
using StockSharp.Algo.Strategies;
using StockSharp.BusinessEntities;
using StockSharp.Messages;

namespace StockSharp.Samples.Strategies;

/// <summary>
/// Polarized Fractal Efficiency strategy.
/// Computes PFE manually from close prices.
/// Buys on PFE turning up from negative, sells on PFE turning down from positive.
/// </summary>
public class PolarizedFractalEfficiencyStrategy : Strategy
{
	private readonly StrategyParam<DataType> _candleType;
	private readonly StrategyParam<int> _pfePeriod;

	private readonly List<decimal> _closes = new();
	private decimal _prevPfe;
	private decimal _prevPrevPfe;
	private int _formed;

	public DataType CandleType { get => _candleType.Value; set => _candleType.Value = value; }
	public int PfePeriod { get => _pfePeriod.Value; set => _pfePeriod.Value = value; }

	public PolarizedFractalEfficiencyStrategy()
	{
		_candleType = Param(nameof(CandleType), TimeSpan.FromMinutes(5).TimeFrame())
			.SetDisplay("Candle Type", "Type of candles", "General");

		_pfePeriod = Param(nameof(PfePeriod), 9)
			.SetDisplay("PFE Period", "Indicator calculation period", "Indicators")
			.SetGreaterThanZero();
	}

	public override IEnumerable<(Security sec, DataType dt)> GetWorkingSecurities()
	{
		return [(Security, CandleType)];
	}

	protected override void OnStarted2(DateTime time)
	{
		base.OnStarted2(time);

		_closes.Clear();
		_prevPfe = 0;
		_prevPrevPfe = 0;
		_formed = 0;

		var sma = new SimpleMovingAverage { Length = 1 };

		var subscription = SubscribeCandles(CandleType);
		subscription.Bind(sma, ProcessCandle).Start();

		var area = CreateChartArea();
		if (area != null)
		{
			DrawCandles(area, subscription);
			DrawOwnTrades(area);
		}
	}

	private void ProcessCandle(ICandleMessage candle, decimal _smaVal)
	{
		if (candle.State != CandleStates.Finished)
			return;

		_closes.Add(candle.ClosePrice);

		var period = PfePeriod;
		if (_closes.Count < period + 1)
			return;

		// Keep only what we need
		while (_closes.Count > period + 2)
			_closes.RemoveAt(0);

		var n = _closes.Count;
		var closeNow = _closes[n - 1];
		var closePast = _closes[n - 1 - period];

		// Direct distance
		var diff = (double)(closeNow - closePast);
		var directDist = Math.Sqrt(diff * diff + (double)(period * period));

		// Sum of bar-to-bar distances
		var sumDist = 0.0;
		for (var i = n - period; i < n; i++)
		{
			var d = (double)(_closes[i] - _closes[i - 1]);
			sumDist += Math.Sqrt(d * d + 1.0);
		}

		if (sumDist == 0)
			return;

		var sign = closeNow >= closePast ? 1.0 : -1.0;
		var pfe = (decimal)(100.0 * sign * directDist / sumDist);

		_formed++;

		if (_formed < 3)
		{
			_prevPrevPfe = _prevPfe;
			_prevPfe = pfe;
			return;
		}

		if (!IsFormedAndOnline())
			return;

		// Trend reversal: PFE was falling and now rising => buy
		if (_prevPfe < _prevPrevPfe && pfe > _prevPfe && Position <= 0)
			BuyMarket();
		// PFE was rising and now falling => sell
		else if (_prevPfe > _prevPrevPfe && pfe < _prevPfe && Position >= 0)
			SellMarket();

		_prevPrevPfe = _prevPfe;
		_prevPfe = pfe;
	}
}

File: API/2316_Polarized_Fractal_Efficiency/PY/polarized_fractal_efficiency_strategy.py
import clr
import math

clr.AddReference("StockSharp.Messages")
clr.AddReference("StockSharp.Algo")

from System import TimeSpan
from StockSharp.Messages import DataType, CandleStates
from StockSharp.Algo.Indicators import SimpleMovingAverage
from StockSharp.Algo.Strategies import Strategy


class polarized_fractal_efficiency_strategy(Strategy):
    def __init__(self):
        super(polarized_fractal_efficiency_strategy, self).__init__()
        self._candle_type = self.Param("CandleType", DataType.TimeFrame(TimeSpan.FromHours(4))) \
            .SetDisplay("Candle Type", "Type of candles", "General")
        self._pfe_period = self.Param("PfePeriod", 9) \
            .SetDisplay("PFE Period", "Indicator calculation period", "Indicators")
        self._closes = []
        self._prev_pfe = 0.0
        self._prev_prev_pfe = 0.0
        self._formed = 0

    @property
    def candle_type(self):
        return self._candle_type.Value

    @property
    def pfe_period(self):
        return self._pfe_period.Value

    def OnReseted(self):
        super(polarized_fractal_efficiency_strategy, self).OnReseted()
        self._closes = []
        self._prev_pfe = 0.0
        self._prev_prev_pfe = 0.0
        self._formed = 0

    def OnStarted(self, time):
        super(polarized_fractal_efficiency_strategy, self).OnStarted(time)
        self._closes = []
        self._prev_pfe = 0.0
        self._prev_prev_pfe = 0.0
        self._formed = 0
        sma = SimpleMovingAverage()
        sma.Length = 1
        subscription = self.SubscribeCandles(self.candle_type)
        subscription.Bind(sma, self.process_candle).Start()
        area = self.CreateChartArea()
        if area is not None:
            self.DrawCandles(area, subscription)
            self.DrawOwnTrades(area)

    def process_candle(self, candle, _sma_val):
        if candle.State != CandleStates.Finished:
            return
        self._closes.append(float(candle.ClosePrice))
        period = int(self.pfe_period)
        if len(self._closes) < period + 1:
            return
        while len(self._closes) > period + 2:
            self._closes.pop(0)
        n = len(self._closes)
        close_now = self._closes[n - 1]
        close_past = self._closes[n - 1 - period]
        diff = close_now - close_past
        direct_dist = math.sqrt(diff * diff + period * period)
        sum_dist = 0.0
        for i in range(n - period, n):
            d = self._closes[i] - self._closes[i - 1]
            sum_dist += math.sqrt(d * d + 1.0)
        if sum_dist == 0:
            return
        sign = 1.0 if close_now >= close_past else -1.0
        pfe = 100.0 * sign * direct_dist / sum_dist
        self._formed += 1
        if self._formed < 3:
            self._prev_prev_pfe = self._prev_pfe
            self._prev_pfe = pfe
            return
        if self._prev_pfe < self._prev_prev_pfe and pfe > self._prev_pfe and self.Position <= 0:
            self.BuyMarket()
        elif self._prev_pfe > self._prev_prev_pfe and pfe < self._prev_pfe and self.Position >= 0:
            self.SellMarket()
        self._prev_prev_pfe = self._prev_pfe
        self._prev_pfe = pfe

    def CreateClone(self):
        return polarized_fractal_efficiency_strategy()
